polkadot/xcm/pallet-xcm/src/ethereum.rs

```diff
@@ -0,0 +1,191 @@
+			}
+		}
+		Ok((fee_transfer_type, assets_transfer_type.ok_or(Error::<T>::Empty)?))
+	}
+
+	fn build_ethereum_xcm_transfer(
+		origin: Location,
+		dest: Location,
+		beneficiary: Location,
+		assets: Vec<Asset>,
+		transfer_type: TransferType,
+		fees: Vec<Asset>,
+		weight_limit: WeightLimit,
+	) -> Result<(Xcm<<T as Config>::RuntimeCall>, Option<Xcm<()>>), Error<T>> {
+		log::debug!(
+			target: "xcm::pallet_xcm::build_xcm_transfer_type",
+			"origin {:?}, dest {:?}, beneficiary {:?}, assets {:?}, transfer_type {:?}, \
+			fees_handling {:?}, weight_limit: {:?}",
+			origin, dest, beneficiary, assets, transfer_type, fees, weight_limit,
+		);
+		let fee_on_bridgehub = &fees[0];
+		let fee_on_ethereum = &fees[1];
+		let (local, mut remote) = match transfer_type {
+			TransferType::LocalReserve => {
+				let (local, remote) = Self::local_reserve_transfer_programs(
+					origin.clone(),
+					dest.clone(),
+					beneficiary,
+					assets,
+					FeesHandling::Batched { fees: fee_on_ethereum.clone() },
+					weight_limit,
+				)?;
+				Some((local, remote))
+			},
+			TransferType::DestinationReserve => {
+				let (local, remote) = Self::destination_reserve_transfer_programs(
+					origin.clone(),
+					dest.clone(),
+					beneficiary,
+					assets,
+					FeesHandling::Batched { fees: fee_on_ethereum.clone() },
+					weight_limit,
+				)?;
+				Some((local, remote))
+			},
+			_ => None,
+		}
+		.ok_or(Error::InvalidAssetUnsupportedReserve)?;
+		remote.inner_mut().push(BurnAsset(fee_on_bridgehub.clone().into()));
+		Ok((local, Some(remote)))
+	}
+
+	fn execute_ethereum_xcm_transfer(
+		origin: Location,
+		dest: Location,
+		fees: Assets,
+		mut local_xcm: Xcm<<T as Config>::RuntimeCall>,
+		remote_xcm: Option<Xcm<()>>,
+	) -> DispatchResult {
+		log::debug!(
+			target: "xcm::pallet_xcm::execute_xcm_transfer",
+			"origin {:?}, dest {:?}, local_xcm {:?}, remote_xcm {:?}",
+			origin, dest, local_xcm, remote_xcm,
+		);
+
+		let weight =
+			T::Weigher::weight(&mut local_xcm).map_err(|()| Error::<T>::UnweighableMessage)?;
+		let mut hash = local_xcm.using_encoded(sp_io::hashing::blake2_256);
+		let outcome = T::XcmExecutor::prepare_and_execute(
+			origin.clone(),
+			local_xcm,
+			&mut hash,
+			weight,
+			weight,
+		);
+		Self::deposit_event(Event::Attempted { outcome: outcome.clone() });
+		outcome.ensure_complete().map_err(|error| {
+			log::error!(
+				target: "xcm::pallet_xcm::execute_xcm_transfer",
+				"XCM execution failed with error {:?}", error
+			);
+			Error::<T>::LocalExecutionIncomplete
+		})?;
+
+		if let Some(remote_xcm) = remote_xcm {
+			let (ticket, _) = validate_send::<T::XcmRouter>(dest.clone(), remote_xcm.clone())
+				.map_err(Error::<T>::from)?;
+			if origin != Here.into_location() {
+				Self::charge_fees(origin.clone(), fees).map_err(|error| {
+					log::error!(
+						target: "xcm::pallet_xcm::execute_xcm_transfer",
+						"Unable to charge fee with error {:?}", error
+					);
+					Error::<T>::FeesNotMet
```

Comment:
`charge_fees` based on the input fee params.

Reply:
this just allows ppl to put nothing here and get away without paying for fees - why not continue to use `price` coming from
```rust
let (ticket, price) = validate_send::<T::XcmRouter>(dest.clone(), remote_xcm.clone()).map_err(Error::<T>::from)?;
```

Reply:
As mentioned above, though it can be sent to BH, it won't pass the check in the outbound router.

Reply:
Sorry a mistake here. The price is the send cost which should be charged, made a fix in https://github.com/Snowfork/polkadot-sdk/pull/135/commits/2a646311d6d736c623e0b5d847fce5a1fb95c1dd